File: typescript/src/types/analytics.ts
/**
 * Type definitions for cross-form analytics
 */

export interface TimelineEvent {
  date: Date;
  type: 'filing' | 'insider_transaction' | 'institutional_change' | 'earnings' | 'announcement';
  formType: string;
  description: string;
  significance: 'low' | 'medium' | 'high';
  details: Record<string, any>;
}

export interface Timeline {
  cik: string;
  companyName: string;
  events: TimelineEvent[];
  period: {
    start: Date;
    end: Date;
  };
}

export interface Correlation {
  type: 'insider_trading_vs_earnings' | 'insider_trading_vs_announcements' | 'institutional_vs_performance';
  strength: number;
  significance: number;
  description: string;
  dataPoints: Array<{
    date: Date;
    value1: number;
    value2: number;
  }>;
}

export interface OwnershipChange {
  period: string;
  institutionalOwnership: {
    previous: number;
    current: number;
    change: number;
  };
  insiderOwnership: {
    previous: number;
    current: number;
    change: number;
  };
  majorShareholders: Array<{
    name: string;
    type: 'institutional' | 'insider' | 'other';
    ownership: number;
    change: number;
  }>;
}

export interface OwnershipTrend {
  cik: string;
  companyName: string;
  changes: OwnershipChange[];
  trends: {
    institutionalTrend: 'increasing' | 'decreasing' | 'stable';
    insiderTrend: 'increasing' | 'decreasing' | 'stable';
    concentration: 'increasing' | 'decreasing' | 'stable';
  };
}

export interface ComplianceMetrics {
  cik: string;
  companyName: string;
  period: {
    start: Date;
    end: Date;
  };
  filingCompliance: {
    onTime: number;
    late: number;
    missed: number;
    score: number;
  };
  insiderCompliance: {
    form4OnTime: number;
    form4Late: number;
    form5Filed: number;
    score: number;
  };
  overallScore: number;
  riskLevel: 'low' | 'medium' | 'high';
}

File: python/src/sec_edgar_toolkit/types/analytics.py
"""Type definitions for cross-form analytics."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, TypedDict


class TimelineEvent(TypedDict):
    """Timeline event information."""

    date: datetime
    type: str  # 'filing', 'insider_transaction', 'institutional_change', 'earnings', 'announcement'
    form_type: str
    description: str
    significance: str  # 'low', 'medium', 'high'
    details: Dict[str, Any]


class TimelinePeriod(TypedDict):
    """Timeline period information."""

    start: datetime
    end: datetime


class Timeline(TypedDict):
    """Company timeline data."""

    cik: str
    company_name: str
    events: List[TimelineEvent]
    period: TimelinePeriod


class DataPoint(TypedDict):
    """Correlation data point."""

    date: datetime
    value1: float
    value2: float


class Correlation(TypedDict):
    """Correlation analysis result."""

    type: str  # 'insider_trading_vs_earnings', 'insider_trading_vs_announcements', 'institutional_vs_performance'
    strength: float
    significance: float
    description: str
    data_points: List[DataPoint]


class OwnershipData(TypedDict):
    """Ownership data for a period."""

    previous: float
    current: float
    change: float


class MajorShareholder(TypedDict):
    """Major shareholder information."""

    name: str
    type: str  # 'institutional', 'insider', 'other'
    ownership: float
    change: float


class OwnershipChange(TypedDict):
    """Ownership change information."""

    period: str
    institutional_ownership: OwnershipData
    insider_ownership: OwnershipData
    major_shareholders: List[MajorShareholder]


class OwnershipTrends(TypedDict):
    """Ownership trend indicators."""

    institutional_trend: str  # 'increasing', 'decreasing', 'stable'
    insider_trend: str  # 'increasing', 'decreasing', 'stable'
    concentration: str  # 'increasing', 'decreasing', 'stable'


class OwnershipTrend(TypedDict):
    """Ownership trend analysis."""

    cik: str
    company_name: str
    changes: List[OwnershipChange]
    trends: OwnershipTrends


class FilingCompliance(TypedDict):
    """Filing compliance metrics."""

    on_time: int
    late: int
    missed: int
    score: float


class InsiderCompliance(TypedDict):
    """Insider compliance metrics."""

    form4_on_time: int
    form4_late: int
    form5_filed: int
    score: float


class CompliancePeriod(TypedDict):
    """Compliance period information."""

    start: datetime
    end: datetime


class ComplianceMetrics(TypedDict):
    """Compliance metrics analysis."""

    cik: str
    company_name: str
    period: CompliancePeriod
    filing_compliance: FilingCompliance
    insider_compliance: InsiderCompliance
    overall_score: float
    risk_level: str  # 'low', 'medium', 'high'
